### PR TITLE
Console/Timeline improvements

### PR DESCRIPTION
**Pull request by shaoshing**
This PR includes the following improvements to the Console and Timeline:

#### Togglable action log
You will be able to disable action log and timestamp completely by turning `Log Actions` and `Log Action Transfers` off. This is useful when you need a clean console or Timeline.

screen shot 2016-02-18 at 3 48 50 pm

#### Better descriptor log
- Descriptor log now shows the params/result in a collapsed group, which avoids overcrowding the console.
- Has stack trace
- Show completion time. This helps target slow descriptors.
- Show timestamp in the Timeline.
- Descriptor calls have unique ID to help mapping the `executing/finished` message in the console, and mapping between the console and timeline.

screen shot 2016-02-18 at 3 49 23 pm

#### Startup timestamp
screen shot 2016-02-18 at 3 48 29 pm

### CODE REVIEW

File: src/index-debug.html [hunk 1]
```diff
@@ -26,6 +26,9 @@
     <meta charset="utf-8">
     <title>Photoshop Design Space (development build)</title>
     <link rel="stylesheet" type="text/css" href="style.css" />
+    <script type="text/javascript">
+        var loadTime = new Date();
```
Comment (iwehrman) [inline src/index-debug.html]: I'm confused about how this ends up in the global scope...?

Comment (iwehrman) [inline src/index-debug.html]: I guess all declarations at the top-level end up in the global scope. Weird. Anyway, I don't like the idea of polluting the global scope. Can we do this locally in `main.startup`?

**Comment by iwehrman** [comment]
One thing I don't like about this is that it's actually frequently very useful for me to see the exactly log of when actions were enqueued, executed and finished. By grouping them like this, it certainly does tidy things up, but also obscures how action execution is interleaved.

File: src/index-debug.html (hunk 1 above)
Comment (shaoshing) [inline src/index-debug.html]: Looks like we can use Chrome's `performance.timing` object to retrieve useful information like the time when browser starts loading the page (https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming). With that, we no longer need the record the load time ourself.

**Comment by shaoshing** [comment]
@iwehrman I've eliminated the use of global `loadTime`, and switch to the descriptor call's real ID. Back to you for reviewing.

File: src/index-debug.html (hunk 1 above)
Comment (shaoshing) [inline src/index-debug.html]: oh, the top-level declarations. They will become global variables, unless in `strict mode`.

**Comment by iwehrman** [comment]
Besides my other comments offline, can we also make action times enabled by default (in debug builds)? This is just to reduce confusion in the short term.

**Comment by iwehrman** [comment]
Also, it doesn't seem like the traces in the descriptor "finished" messages are ever very interesting. But maybe I've just been missing the interesting cases.

**Comment by iwehrman** [comment]
Also, it might be possible to format these messages a little differently to make it easier to match up the calls. For example, instead of:

```
[Descriptor] Executing get - 1544
[Descriptor] Finished get in 21ms - 1544
```
it might help to align the IDs and stick the time at the end:
```
[Descriptor] Executing get - 1544
[Descriptor] Finished get - 1544 in 21ms
```

And, as mentioned offline, red highlighting for long calls would be awesome :)

**Comment by shaoshing** [comment]
@iwehrman below is the updated version

- highlight slow descriptor (> 100ms)
- remove trace for the `finished` log
- move time to the end
- show summarized commands for batchPlay
- Enable `Log Actions` by default

screen shot 2016-02-19 at 12 46 37 pm

**Comment by shaoshing** [comment]
Fixed the repeated `in` in the message

File: src/js/fluxcontroller.js [hunk 2]
```diff
@@ -781,7 +783,8 @@ define(function (require, exports, module) {
     FluxController.prototype._applyAction = function (action, actionReceiver, params, parentActionName) {
         var flux = this.flux,
             actionObject = action.action,
-            actionName = this._actionNames.get(action);
+            actionName = this._actionNames.get(action),
+            logActions = __PG_DEBUG__ && this.flux.store("preferences").getState().get("logActions") !== false;
```
Comment (iwehrman) [inline src/js/fluxcontroller.js]: A slightly nicer way to do this is `.get("logActions", true)`.

File: src/js/actions/debug.js [hunk 3]
```diff
@@ -374,43 +400,124 @@ define(function (require, exports) {
...
+                return id;
+            }, {
+                context: window._spaces.ps.descriptor,
+                multiArgs: true
+            });
 
-                    return result;
+            var summarizeCommands = function (commands) {
+                if (commands.length === 0) {
+                    return "EMPTY";
```
Comment (iwehrman) [inline src/js/actions/debug.js]: Can we just make this the empty string here so this shows up as `Executing batchPlay: [] - 123`?

**Comment by iwehrman** [comment]
Well, this is great. Two more micro nits and then I'll merge.

**Comment by shaoshing** [comment]
Nits addressed! Back to you @iwehrman

**Comment by iwehrman** [comment]
Super cool!

**Pull request merged**